src/audio/detect_test.c

```diff
@@ -0,0 +1,654 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Slawomir Blauciak <slawomir.blauciak@linux.intel.com>
+ */
+
+#include <stdint.h>
+#include <stddef.h>
+#include <sof/lock.h>
+#include <sof/list.h>
+#include <sof/stream.h>
+#include <sof/ipc.h>
+#include <sof/notifier.h>
+#include <sof/audio/component.h>
+#include <sof/audio/kpb.h>
+#include <uapi/user/detect_test.h>
+
+/* tracing */
+#define trace_keyword(__e, ...) \
+	trace_event(TRACE_CLASS_KEYWORD, __e, ##__VA_ARGS__)
+#define trace_keyword_error(__e, ...) \
+	trace_error(TRACE_CLASS_KEYWORD, __e, ##__VA_ARGS__)
+#define tracev_keyword(__e, ...) \
+	tracev_event(TRACE_CLASS_KEYWORD, __e, ##__VA_ARGS__)
+
+#define ACTIVATION_DEFAULT_SHIFT 3
+#define ACTIVATION_DEFAULT_THRESHOLD 0.5
+
+#define ACTIVATION_DEFAULT_THRESHOLD_S16 \
+	((int16_t)((INT16_MAX) * (ACTIVATION_DEFAULT_THRESHOLD)))
+
+/* number of samples to be treated as a full keyphrase */
+#define KEYPHRASE_DEFAULT_PREAMBLE_LENGTH (30 * 1024)
+
+struct comp_data {
+	uint32_t period_bytes;		/**< source number of period bytes */
+
+	struct sof_detect_test_config config;
+	void *load_memory;	/**< synthetic memory load */
+	int16_t activation;
+	uint32_t detected;
+	uint32_t detect_preamble; /**< current keyphrase preamble length */
+	uint32_t keyphrase_samples; /**< keyphrase length in samples */
+	uint32_t buf_copy_pos; /**< current copy position for incoming data */
+
+	struct notify_data event;
+	struct kpb_event_data event_data;
+	struct kpb_client client_data;
+
+	void (*detect_func)(struct comp_dev *dev,
+			    struct comp_buffer *source, uint32_t frames);
+};
+
+static void notify_host(struct comp_dev *dev)
+{
+	struct sof_ipc_comp_event event;
+
+	trace_keyword("notify_host()");
+
+	event.event_type = SOF_CTRL_EVENT_KD;
+	event.num_elems = 0;
+
+	ipc_send_comp_notification(dev, &event);
+}
+
+static void notify_kbd(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	trace_keyword("notify_kbd(), preamble: %u", cd->detect_preamble);
+
+	cd->client_data.r_ptr = NULL;
+	cd->client_data.sink = NULL;
+	cd->client_data.id = 0; /**< TODO: acquire proper id from kpb */
+	cd->client_data.history_end = 0; /**< keyphrase end, 0 is now */
+	cd->client_data.history_begin = cd->detect_preamble;
+	/* two channels */
+	cd->client_data.history_depth = cd->detect_preamble *
```

Comment:
Let's change it to milliseconds instead of bytes - it will be more scalable and format independent.

Reply:
I'd rather not modify the logic behind kpb in this PR.

Reply:
As per our discussion earlier, changing as you suggested.